// @mrgnlabs/mrgn-common/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JUPITER_V6_PROGRAM = exports.LUT_PROGRAM_AUTHORITY_INDEX = exports.HOURS_PER_YEAR = exports.LST_MINT = exports.WSOL_MINT = exports.USDC_DECIMALS = exports.DEFAULT_CONFIRM_OPTS = exports.DEFAULT_SEND_OPTS = exports.DEFAULT_COMMITMENT = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.DEFAULT_COMMITMENT = "processed";
exports.DEFAULT_SEND_OPTS = {
    skipPreflight: false,
    preflightCommitment: exports.DEFAULT_COMMITMENT,
};
exports.DEFAULT_CONFIRM_OPTS = {
    commitment: exports.DEFAULT_COMMITMENT,
    ...exports.DEFAULT_SEND_OPTS,
};
exports.USDC_DECIMALS = 6;
exports.WSOL_MINT = new web3_js_1.PublicKey("So11111111111111111111111111111111111111112");
exports.LST_MINT = new web3_js_1.PublicKey("LSTxxxnJzKDFSLr4dUkPcmCf5VyryEqzPLz5j4bpxFp");
exports.HOURS_PER_YEAR = 365.25 * 24;
exports.LUT_PROGRAM_AUTHORITY_INDEX = 5;
exports.JUPITER_V6_PROGRAM = new web3_js_1.PublicKey("JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4");
